docs/2019-05-23-Unification-Modulo-Equations.md

```diff
@@ -0,0 +1,159 @@
+in a principled way without changing the backend for each of these cases,
+and to easily generate proofs if/when that will be desirable.
+
+Background
+----------
+
+The most general unifier of `φ` and `ψ` is `⌈φ∧ψ⌉`
+
+There are some unification cases that are easy to solve, e.g. if `C` is a
+constructor (actually, it only needs to be injective), then
+```
+⌈C(φ_1, …, φ_n) ∧ C(ψ_1, …, ψ_n)⌉=⌈φ_1∧ψ_1⌉∧…∧⌈φ_n∧ψ_n⌉
+```
+
+However, if `f` and `g` are random symbols, without any properties, then
+`⌈f(φ_1, …, φ_n) ∧ g(ψ_1, …, ψ_m)⌉` cannot be reduced to a nicer form.
+
+If `f` is, say, an associative binary operation with neutral element (i.e. the
+concatenation for a list), then we have a reasonable intuition about how to
+unify it, but I don’t know of any written text about that.
+
+Problem
+-------
+
+We are trying to extend unification to sorts that do not have constructors, but
+which have something resembling constructors. An example is a list, which has:
+`unit` - the empty list.
+`elem` - takes an element of a different sort and produces a single element list.
+`concat` - concatenates two lists.
+Heating and cooling could be another example.
+
+Proposed Solution
+-----------------
+
+The first attempt to solve this would be to define some sort of
+no-junk/no-confusion axiom
+for that sort (e.g. any list can be built with `unit`, `elem` and `concat`),
+and then expand the two terms in all possible ways using the available
+equations, then match the two terms at the top as long as we stay within
+the same sort, then use unification for the remainders.
+
+However, this has a few problems. First, there is no way of proving that
+this is the most general unifier, i.e. there is no way of showing that we
+truly expanded the term in all possible ways. Second, this would be really
+inefficient.
+
+### Actual proposal
+
+The second attempt, however, would be to define some unification axioms (I
+think that they are, indeed, axioms, but they could be theorems if we find
+a way to infer them), specifying how to solve certain unification cases
+in a way that can be used almost directly by the Haskell backend (some
+changes needed).
+
+These axioms would allow us to reduce the hard to handle
+unification cases to unambiguous ones (e.g. reduce AC unification to
+constructor-based unification).
+
+See the use-cases below for examples and explanations.
+
+We could write these in a way that we don't duplicate unification cases.
+Then we could apply these as simplification axioms, defining our unification
+strategy in k. When we need performance, we could implement the ones that are
+used more frequently in Haskell.
+
+Or, if that ever becomes desirable (e.g. if we want it to make proof
+certificates easier to generate), we could give up on performance and define
+all simplification cases as axioms/theorems.
+
+Use-case: lists
+---------------
+
+For lists, we could have the following axioms
```

Comment:
I like this proposal! I have a lot of questions about how we should implement it. Maybe these are premature, but I will ask them anyway. How would we apply these axioms? By matching on `⌈_∧_⌉`? Should we use Kore's `\in` connective instead, so we do not just apply any random axiom which has a similar form? Should we have a `unification` attribute? Are the variables in the equations element or set variables? It seems to me like we could have them be set variables to avoid generating lots of extra `\ceil`s.

Reply:
Here are my preferences: apply them by matching, but we should normalize the terms before; use an `unification` attribute.

Yes, we should certainly use set variables! That also came up during a discussion with @traiansf . I've updated the document to reflect that.